shell/platform/android/test/io/flutter/embedding/engine/dynamicfeatures/PlayStoreDynamicFeatureManagerTest.java

```diff
@@ -0,0 +1,143 @@
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.anyInt;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
+
+import android.content.Context;
+import android.content.pm.PackageManager.NameNotFoundException;
+import android.content.res.AssetManager;
+import androidx.annotation.NonNull;
+import io.flutter.embedding.engine.FlutterJNI;
+import java.io.File;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.RuntimeEnvironment;
+import org.robolectric.annotation.Config;
+
+@Config(manifest = Config.NONE)
+@RunWith(RobolectricTestRunner.class)
+public class PlayStoreDynamicFeatureManagerTest {
+  private class TestFlutterJNI extends FlutterJNI {
+    public int loadDartDeferredLibraryCalled = 0;
+    public int updateAssetManagerCalled = 0;
+    public int dynamicFeatureInstallFailureCalled = 0;
+    public String[] searchPaths;
+    public int loadingUnitId;
+
+    public TestFlutterJNI() {}
+
+    @Override
+    public void loadDartDeferredLibrary(int loadingUnitId, @NonNull String[] searchPaths) {
+      loadDartDeferredLibraryCalled++;
+      this.searchPaths = searchPaths;
+      this.loadingUnitId = loadingUnitId;
+    }
+
+    @Override
+    public void updateAssetManager(
+        @NonNull AssetManager assetManager, @NonNull String assetBundlePath) {
+      updateAssetManagerCalled++;
+      this.loadingUnitId = loadingUnitId;
+    }
+
+    @Override
+    public void dynamicFeatureInstallFailure(
+        int loadingUnitId, @NonNull String error, boolean isTransient) {
+      dynamicFeatureInstallFailureCalled++;
+    }
+  }
+
+  // Skips the download process to directly call the loadAssets and loadDartLibrary methods.
+  private class TestPlayStoreDynamicFeatureManager extends PlayStoreDynamicFeatureManager {
+    public TestPlayStoreDynamicFeatureManager(Context context, FlutterJNI jni) {
+      super(context, jni);
+      jni.setDynamicFeatureManager(this);
+    }
+
+    @Override
+    public void downloadDynamicFeature(int loadingUnitId, String moduleName) {
+      // Override this to skip the online SplitInstallManager portion.
+      loadAssets(loadingUnitId, moduleName);
+      loadDartLibrary(loadingUnitId, moduleName);
+    }
+  }
+
+  @Test
+  public void downloadCallsJNIFunctions() throws NameNotFoundException {
+    TestFlutterJNI jni = new TestFlutterJNI();
+    Context spyContext = spy(RuntimeEnvironment.systemContext);
+    doReturn(spyContext).when(spyContext).createPackageContext(any(), anyInt());
+    doReturn(null).when(spyContext).getAssets();
+    String soTestPath = "test/path/app.so-123.part.so";
+    doReturn(new File(soTestPath)).when(spyContext).getFilesDir();
+    TestPlayStoreDynamicFeatureManager playStoreManager =
+        new TestPlayStoreDynamicFeatureManager(spyContext, jni);
+    jni.setDynamicFeatureManager(playStoreManager);
```

Comment:
you're doing this twice. Though I think it's better to do it in the test explicitly than in the constructor